R/utils.R

```diff
@@ -0,0 +1,325 @@
+# Helper functions for do_read_bib()
+
+#' Parse (at)string keys
+#'
+#' Returns a table with all the string values and their value to be used on the
+#' `bib` file.
+#'
+#' @noRd
+parse_strings <- function(map_string,
+                          stringlines) {
+
+  # This needs to be done in a loop, since strings can depend also
+  # on values of previous strings
+
+  parsed_string <- NULL
+
+  for (i in seq_len(length(stringlines))) {
+    row <- map_string[i, ]
+    singleline <- stringlines[i]
+
+    map_braces <- check_balanced_braces(singleline, row$init)
+
+    # Extract string between first and last brace
+
+    f <- map_braces$pos[1]
+    l <- min(map_braces[map_braces$cum_flag == 0, ]$pos)
+
+    singleline <- trimws(substr(singleline, f, l))
+
+    # Remove also the braces
+    singleline <- trimws(gsub("^\\{|\\}$", "", singleline))
+
+    # Split on name/value
+    splitted <- unlist(strsplit(singleline, "="))
+
+    finalrow <- data.frame(
+      name_string = trimws(splitted[1]),
+      value = trimws(splitted[-1])
+    )
+
+    # As character to avoid compatibility issues with older R versions
+    finalrow$name_string <- as.character(finalrow$name_string)
+    finalrow$value <- as.character(finalrow$value)
+
+    parsed_string <- rbind(parsed_string, finalrow)
+  }
+
+  # Need to replace and concat values on loop
+  # Iterate over map_parsed and concat
+  init <- parsed_string[1, ]
+
+  # Trim leading and trailing commas
+  init$value <- trimws(init$value, "both")
+  init$value <- trimws(init$value, "both", whitespace = "\"")
+
+  # If only one string return
+  if (nrow(parsed_string) == 1) {
+    return(init)
+  }
+  for (j in seq(2, nrow(parsed_string))) {
+
+    # Search if need to replace and concat
+    row <- parsed_string[j, ]
+
+    value <- replace_string_and_concat(
+      row$value,
+      init$name_string,
+      init$value
+    )
+
+    row$value <- trimws(value, "both", "\"")
+
+    init <- rbind(init, row)
+  }
+
+  return(init)
+}
+
+#' Parse a single entry
+#'
+#' Returns an entry from a .bib file parsed into an R object
+#'
```

Comment:
Please add some documentation on parameters. (Even though it's an internal function.)

Reply:
Same, this is now documented and with examples

Reply:
And the examples are also part of the tests